http_signer.go

```diff
@@ -0,0 +1,124 @@
+package tests
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"io"
+	"io/ioutil"
+	"math/rand"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	. "github.com/smartystreets/goconvey/convey"
+
+	"github.com/beyondstorage/go-storage/v4/pkg/randbytes"
+	"github.com/beyondstorage/go-storage/v4/types"
+)
+
+func TestHTTPSignerRead(t *testing.T, store types.Storager) {
+	Convey("Given a basic Storager", t, func() {
+		signer, ok := store.(types.HTTPSigner)
+		So(ok, ShouldBeTrue)
+
+		Convey("When read via QuerySignHTTP", func() {
+			size := rand.Int63n(4 * 1024 * 1024)
+			content, err := ioutil.ReadAll(io.LimitReader(randbytes.NewRand(), size))
+			if err != nil {
+				t.Error(err)
+			}
+
+			path := uuid.New().String()
+			_, err = store.Write(path, bytes.NewReader(content), size)
+			if err != nil {
+				t.Error(err)
+			}
+			defer func() {
+				err := store.Delete(path)
+				if err != nil {
+					t.Error(err)
+				}
+			}()
+
+			req, err := signer.QuerySignHTTP(types.OpStoragerRead, path, time.Duration(time.Hour))
+
+			Convey("The error should be nil", func() {
+				So(err, ShouldBeNil)
+
+				So(req, ShouldNotBeNil)
+				So(req.URL, ShouldNotBeNil)
+			})
+
+			client := http.Client{}
+			resp, err := client.Do(req)
+			Convey("The request returned error should be nil", func() {
+				So(err, ShouldBeNil)
+				So(resp, ShouldNotBeNil)
+			})
+
+			defer resp.Body.Close()
+
+			buf, err := ioutil.ReadAll(resp.Body)
+			Convey("The content should be match", func() {
+				So(err, ShouldBeNil)
+				So(buf, ShouldNotBeNil)
+
+				So(resp.ContentLength, ShouldEqual, size)
+				So(sha256.Sum256(buf), ShouldResemble, sha256.Sum256(content))
+			})
+		})
+	})
+}
+
+func TestHTTPSignerWrite(t *testing.T, store types.Storager) {
+	Convey("Given a basic Storager", t, func() {
+		signer, ok := store.(types.HTTPSigner)
+		So(ok, ShouldBeTrue)
+
+		Convey("When write via QuerySignHTTP", func() {
+			size := rand.Int63n(4 * 1024 * 1024)
+			content, err := ioutil.ReadAll(io.LimitReader(randbytes.NewRand(), size))
+			if err != nil {
+				t.Error(err)
+			}
+
+			path := uuid.New().String()
+			req, err := signer.QuerySignHTTP(types.OpStoragerWrite, path, time.Duration(time.Hour))
```

Comment:
Do we need to set `size` here?

I think we need a demo to make sure our test logic is correct. Maybe in s3?

Reply:
> Do we need to set `size` here?

Dose it mean set `Content-Length`? I tried to add `req.Header.Add("Content-Length", strconv.FormatInt(size,10))` and tested with qingstor, the request was successful, but the request is chunked transfer encoding. So I removed the content-length setting.

Reply:
According to the comment for `Header` in `http.Request`, for client requests, certain headers such as Content-Length and Connection are automatically written when needed and values in Header may be ignored.

Reply:
> I think we need a demo to make sure our test logic is correct. Maybe in s3?

I'll privde a demo.

Reply:
> chunked transfer encoding

Strange, chunked transfer encoding should be used for requests that don't have `content-length` header.

Let's discuss the detail in the demo of s3 instead.